KubeArmor/feeder/policyMatcher.go

```diff
@@ -12,450 +13,348 @@ import (
 // == Security Policies == //
 // ======================= //
 
-// UpdateSecurityPolicies Function
-func (fd *Feeder) UpdateSecurityPolicies(action string, conGroup tp.ContainerGroup) {
-	if action == "DELETED" {
-		delete(fd.SecurityPolicies, conGroup.NamespaceName+"_"+conGroup.ContainerGroupName)
-	} else { // ADDED | MODIFIED
-		matches := tp.MatchPolicies{}
-
-		for _, secPolicy := range conGroup.SecurityPolicies {
-			if len(secPolicy.Spec.AppArmor) > 0 {
-				match := tp.MatchPolicy{}
-
-				match.PolicyName = secPolicy.Metadata["policyName"]
-				match.Native = true
-
-				matches.Policies = append(matches.Policies, match)
-				continue
-			}
-
-			if len(secPolicy.Spec.Process.MatchPaths) > 0 {
-				for _, path := range secPolicy.Spec.Process.MatchPaths {
-					if len(path.FromSource) == 0 {
-						match := tp.MatchPolicy{}
-
-						match.PolicyName = secPolicy.Metadata["policyName"]
-
-						match.Severity = strconv.Itoa(path.Severity)
-						match.Tags = path.Tags
-						match.Message = path.Message
-
-						match.Source = ""
-						match.Operation = "Process"
-						match.Resource = path.Path
-
-						match.Action = path.Action
-
-						matches.Policies = append(matches.Policies, match)
-					} else {
-						for _, src := range path.FromSource {
-							if len(src.Path) > 0 {
-								match := tp.MatchPolicy{}
-
-								match.PolicyName = secPolicy.Metadata["policyName"]
-
-								match.Severity = strconv.Itoa(path.Severity)
-								match.Tags = path.Tags
-								match.Message = path.Message
-
-								match.Source = src.Path
-								match.Operation = "Process"
-								match.Resource = path.Path
+func newProcPathMatchPolicy(policyName, source string, path tp.ProcessPathType) tp.MatchPolicy {
+	match := tp.MatchPolicy{
+		PolicyName: policyName,
+		Severity:   strconv.Itoa(path.Severity),
+		Tags:       path.Tags,
+		Message:    path.Message,
+		Source:     source,
+		Operation:  "Process",
+		Resource:   path.Path,
+		Action:     path.Action,
+	}
```

Comment:
These are just helper functions. It could be managed through interfaces instead of helpers, but as a first proposal I preferred not to touch types.go concerning that. Another additional discussion would be related to the use of embedded structure/interface types. I noticed that some have similar structure.

Reply:
@nam-jaehyun this can be changed for:

```golang
func newMatchPolicy(policyName, src, op string, mp interface{}) tp.MatchPolicy {
	match := tp.MatchPolicy{
		PolicyName: policyName,
		Source:     src,
		Operation:  op,
	}

	if ppt, ok := mp.(tp.ProcessPathType); ok {
		match.Severity = strconv.Itoa(ppt.Severity)
		match.Tags = ppt.Tags
		match.Message = ppt.Message
		match.Resource = ppt.Path
		match.Action = ppt.Action
	} else if pdt, ok := mp.(tp.ProcessDirectoryType); ok {
		match.Severity = strconv.Itoa(pdt.Severity)
		match.Tags = pdt.Tags
		match.Message = pdt.Message
		match.Resource = pdt.Directory
		match.Action = pdt.Action
	} else if fpt, ok := mp.(tp.FilePathType); ok {
		match.Severity = strconv.Itoa(fpt.Severity)
		match.Tags = fpt.Tags
		match.Message = fpt.Message
		match.Resource = fpt.Path
		match.Action = fpt.Action
	} else if fdt, ok := mp.(tp.FileDirectoryType); ok {
		match.Severity = strconv.Itoa(fdt.Severity)
		match.Tags = fdt.Tags
		match.Message = fdt.Message
		match.Resource = fdt.Directory
		match.Action = fdt.Action
	} else if npt, ok := mp.(tp.NetworkProtocolType); ok {
		match.Severity = strconv.Itoa(npt.Severity)
		match.Tags = npt.Tags
		match.Message = npt.Message
		match.Resource = getProtocolFromName(npt.Protocol)
		match.Action = npt.Action
	} else if cct, ok := mp.(tp.CapabilitiesCapabilityType); ok {
		match.Severity = strconv.Itoa(cct.Severity)
		match.Tags = cct.Tags
		match.Message = cct.Message
		match.Resource = getCapabilityFromName(cct.Capability)
		match.Action = cct.Action
	} else {
               // type not found
        }

	return match
}
```

It could be smaller but we lack of common structures/interfaces.